keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,257 @@
+kind: ServicePolicy
+metadata:
+  name: service-policy-example-1
+  namespace: foo
+spec:
+  serviceSelector:
+    matchLabels:
+      app: bar
+  topology:
+    key: kubernetes.io/region
+    mode: required    
+---
+kind: ServicePolicy
+metadata:
+  name: service-policy-example-2
+  namespace: foo
+spec:
+  serviceSelector:
+    matchLabels:
+      app: bar
+  topology:
+    key: kubernetes.io/switch
+    mode: required   
+```
+
+In our example, services in namespace `foo` with label `app=bar` will be dominated by both `service-policy-example-1` and `service-policy-example-2`. Requests to these services will be routed only to backends that satisfy both same region and same switch as kube-proxy.
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname.
+
+So, create a new `Topology` field in `Endpoints.Subsets.Addresses` for identifying what topology domain the endpoints pod exists, e.g. what host, rack, zone, region etc. In other words, copy the topology-related labels of node hosting the endpoint to `EndpointAddress.Topology`.
+
+```go
+type EndpointAddress struct {
+  // labels of node hosting the endpoint
+  Topology map[string]string
+}
+```
+
+### Endpoints Controller changes
+
+Endpoint Controller will populate the `Topology` for each `EndpointAddress`. We want `EndpointAddress.Topology` to tell the LB, such as kube-proxy what topological domain(e.g. host, rack, zone, region etc.) the endpoints is in.
+
+Endpoints controller will need to watch two extra resources: ServicePolicy and Nodes. Watching ServicePolicy for knowing what services have topological preferences. Watching Nodes for knowing labels of node hosting the endpoint and copy the node labels referenced in the service spec's topology constraints to EndpointAddress.
+
+Endpoints Controller will maintain two extra caches: `NodeToPodsCache` and `ServiceToPoliciesCache`. 
+`NodeToPodsCache` maps the node's name to the pods running on it. Node's add, delete and labels' change will trigger `NodeToPodsCache` reindex.
+
+`ServiceToPoliciesCache` maps the Service's namespaced name to all of its ServicePolicys.
+
+So, the new logic of endpoint controller might like:
+
+```go
+go watch Node, ServicePolicy
+// In each sync loop, given a service, sync its endpoints
+for i, pod := range service backends; do
+  servicePolicys := ServiceToPoliciesCache[service.Name]
+  node := nodeCache[pod.Spec.NodeName]
+  // endpointAddress := &v1.EndpointAddress {}
+  // Copy all topology-related labels of node hosting endpoint to endpoint
+  // We can only include node labels referenced in the service spec's topology constraints
+  for _, servicePolicy := range servicePolicys; do
+    topoKey := servicePolicy.Topology.Key
+    endpointAddress.Topology[topoKey] = node.Labels[topoKey]
+  done
+  endpoints.Subsets[i].Addresses = endpointAddress
+done
+```
+
+### Kube-proxy changes
```

Comment:
Should this impact kube-dns as well?

Reply:
Good question - it would be somewhat complex for DNS to do split-horizon responses based on client IP, but that is the implication of this (for headless services, anyway). @johnbelamaric

Reply:
It requires watching pods which isn't something we would want to do in general, and not something we would want to adapt kube-dns to do. CoreDNS can optionally do that already so if the customer is willing to accept the cost of that we can do it for headless.

For purposes of this KEP we can make is clusterIP services only, and then the headless version can become a feature of CoreDNS maybe. If we do that though, should we enforce anything in the API if the user configures topology for a headless service (assuming we go the route of adding this to service rather than a separate resource)? Otherwise it could cause expectations that aren't really being met.

Reply:
I think I missed the headless service part in the proposal, will take a deep look at the effect.